### PR TITLE
feat(Form): improvements to be able to build custom fields

### PR DESCRIPTION
**Pull request by sergivillar**

### CODE REVIEW

File: src/__stories__/custom-input-story.tsx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+} from '..';
+import {DOWN, ENTER, SPACE, UP} from '../utils/key-codes';
+import {cancelEvent} from '../utils/dom';
+import {combineRefs} from '../utils/common';
+
+export default {
+    title: 'Components/Forms/Custom input',
+};
+
+const OPTIONS = [
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:25]: Make this a list of countries or fruits or whatever... If all options starts with "Option" it always autocompletes all or nothing.

I would be also interesting for this example to have scroll in the list of suggestions

File: src/__stories__/custom-input-story.tsx [hunk 2]
```diff
@@ -0,0 +1,269 @@
+import * as React from 'react';
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:1]: I find this annoying
https://user-images.githubusercontent.com/579417/149164205-cc34a187-9149-4b69-aa49-212fcbbf756a.mp4
Not sure if it is easy to fix. If you click outside the menu to close it, change the tab and go back to it, the menu opens automatically

File: src/__stories__/custom-input-story.tsx [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    createUseStyles,
+    Form,
+    Stack,
+} from '..';
+import {DOWN, ENTER, SPACE, UP} from '../utils/key-codes';
+import {cancelEvent} from '../utils/dom';
+import {combineRefs} from '../utils/common';
+
+export default {
+    title: 'Components/Forms/Custom input',
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:22]: Please, rename to "Custom field"

File: src/__stories__/custom-input-story.tsx [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            setSelectedValue('');
+        }
+    };
+
+    React.useEffect(() => {
+        setOptions(OPTIONS.filter(({value}) => value.includes(filterValue)));
+    }, [filterValue]);
+
+    return (
+        <StorySection title="Custom input">
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:225]: Custom field

File: src/__stories__/custom-input-story.tsx [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            <Stack space={32}>
+                <Text2 regular>
+                    This is an example of how to create a custom form input based on Mistica's pieces. This
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:228]: custom form field

File: src/__stories__/custom-input-story.tsx [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                                options={options}
+                                className={className}
+                                selectedValue={selectedValue}
+                                setFilterValue={setFilterValue}
+                                setSelectedValue={setSelectedValue}
+                                closeMenu={close}
+                            />
+                        )}
+                    />
+                </Form>
```
Comment (pladaria) [inline src/__stories__/custom-input-story.tsx:263]: Better extract this into a standalone component, for example `AutocompleteSelectField`

And, to complete the example, expose `useFieldProps` and use that hook in this component. Add a submit button and check that the component is registered and working as expected.

File: src/__stories__/custom-input-story.tsx (hunk 1 above)
Comment (sergivillar) [reply src/__stories__/custom-input-story.tsx:25]: done

**Review by pladaria — APPROVED** [review]

File: src/__stories__/custom-field-story.tsx [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        error: false,
+        disabled: false,
+        onBlur: undefined,
+        validate: undefined,
+        onChange: undefined,
+        onChangeValue: undefined,
+    });
+
+    React.useEffect(() => {
+        setFilteredOptions(options.filter(({value}) => value.includes(filterValue)));
```
Comment (atabel) [inline src/__stories__/custom-field-story.tsx:235]: why don't you filter at render time?

**Review by atabel — APPROVED** [review]

File: src/__stories__/custom-field-story.tsx (hunk 7 above)
Comment (sergivillar) [reply src/__stories__/custom-field-story.tsx]: changed to filter on onChangeValue

**Pull request merged**